default/python/AI/freeorion_tools/_freeorion_tools.py

```diff
@@ -349,3 +349,31 @@ def get_ship_part(part_name: str):
         warning("Could not find part %s" % part_name)
 
     return part_type
+
+
+def named_int(name: str) -> int:
+    """
+    Returns a NamedReal from FOCS.
+    If the value does not exist, reports an error and returns 1.
+    Note that we do not raise and exception so that the AI can continue, as good as it can, with outdated information.
+    This is also why we return 1, returning 0 could cause followup errors if the value is used as divisor.
+    """
+    value = fo.getNamedInt(name)
+    if value is None:
+        error(f"Requested NamedInt {name}, which doesn't exist!")
+        value = 1
+    return value
+
+
+def named_real(name: str) -> float:
+    """
+    Returns a NamedReal from FOCS.
+    If the value does not exist, reports an error and returns 1.0.
+    Note that we do not raise and exception so that the AI can continue, as good as it can, with outdated information.
+    This is also why we return 1, returning 0 could cause followup errors if the value is used as divisor.
```

Comment:
arguably better to add safety checks on the divisor wherever that error could happen...?

Reply:
The only case where I have done it so far was dividing by BUREAUCRACY_ADMINS_NEEDED_TURN_SCALING, which hopefully is gone soon anyway. This value was also used as a divisor in FOCS itself.

In the end, when a value that is used by the AI is removed from FOCS, the AI should be adopted. But in doubt using a wrong value should still be better than throwing an exception, which means skipping all the later code for that turn, although most of it may still work correctly.

Reply:
the point wouldn't be to throw an exception, but rather to check that the value that is returned from the API call won't cause an exception to be thrown, which could occur if the referenced named value is removed and as a result the returned value is different from what the AI code expected

Reply:
That point is beyond the scope of this PR. I've chosen a default that is less likely to cause problems, while the fact that the default is used is reported by an error. There is currently not a single case where a named value is used as a divisor.